dbt/include/snowflake/macros/utils/safe_cast.sql

```diff
@@ -1,3 +1,13 @@
 {% macro snowflake__safe_cast(field, type) %}
-    try_cast({{field}} as {{type}})
+    {% if type|upper == "GEOMETRY" -%}
+        try_to_geometry({{field}})
+    {% elif type|upper == "GEOGRAPHY" -%}
+        try_to_geography({{field}})
+    {% elif type|upper != "VARIANT" -%}
```

Comment:
In the case where the field type _is_ `variant` or `object` or `array`, I recall having to use `try_parse_json`. Otherwise I get errors like:
```
Database Error
    001007 (22023): SQL compilation error:
    invalid type [TRY_CAST('{''Alberta'':''Edmonton'',''Manitoba'':''Winnipeg''}')] for parameter 'TO_OBJECT'
```
```
Database Error
    001065 (22023): SQL compilation error:
    Function TRY_CAST cannot be used with arguments of types VARCHAR(13) and ARRAY
```

Reply:
Could you share your tophatting setup?

The tests setup is: 

object type: 

```sql
--models/my_upstream_model.sql 

select
  {'Alberta':'Edmonton','Manitoba':'Winnipeg'} as tested_column
```

```sql
--models/my_model.sql 

select
    tested_column from {{ ref('my_upstream_model')}}
```

```sql
--models/schema.yml 

unit_tests:
  - name: test_my_model
    model: my_model
    given:
      - input: ref('my_upstream_model')
        rows:
          - { tested_column: {'Alberta':'Edmonton','Manitoba':'Winnipeg'} }
    expect:
      rows:
        - { tested_column: {'Alberta':'Edmonton','Manitoba':'Winnipeg'} }
```

& the generated sql is
```sql
with dbt_internal_unit_test_actual as (
  select
    tested_column, 'actual' as "actual_or_expected"
  from (
    with __dbt__cte__my_upstream_model as (

-- Fixture for my_upstream_model
select 
    
        try_cast({'Alberta': 'Edmonton', 'Manitoba': 'Winnipeg'} as OBJECT)
     as tested_column
) select
    tested_column from __dbt__cte__my_upstream_model
  ) _dbt_internal_unit_test_actual
),
-- Build expected result
dbt_internal_unit_test_expected as (
  select
    tested_column, 'expected' as "actual_or_expected"
  from (
    select {'Alberta': 'Edmonton', 'Manitoba': 'Winnipeg'} as tested_column
  ) _dbt_internal_unit_test_expected
)
-- Union actual and expected results
select * from dbt_internal_unit_test_actual
union all
select * from dbt_internal_unit_test_expected
```

Reply:
Ah! I had it specified with an extra set of quotes wrapping the dictionaries (i.e. as strings):
```yaml
unit_tests:
  - name: test_my_model
    model: model_b
    given:
      - input: ref('model_a')
        rows:
          - {my_field: :"{'Alberta':'Edmonton','Manitoba':'Winnipeg'}}
    expect:
      rows:
        - {my_field: "{'Alberta':'Edmonton','Manitoba':'Winnipeg'}"}
```
When I removed the quotes, it just works. Meaning this also just works:
```yaml
unit_tests:
  - name: test_my_model
    model: model_b
    given:
      - input: ref('model_a')
        rows:
          - my_field:
              Alberta: Edmonton
              Manitoba: Winnipeg
    expect:
      rows:
        - my_field:
            Alberta: Edmonton
            Manitoba: Winnipeg
```

That's really very cool. Nice work!!